network/src/commonMain/kotlin/com/wire/kalium/network/utils/NetworkUtils.kt

```diff
@@ -0,0 +1,54 @@
+package com.wire.kalium.network.utils
+
+import com.wire.kalium.network.api.ErrorResponse
+import com.wire.kalium.network.exceptions.KaliumException
+import io.ktor.client.call.*
+import io.ktor.client.features.*
+import io.ktor.client.statement.*
+import io.ktor.util.*
+import kotlinx.serialization.SerializationException
+
+sealed class NetworkResponse<out T> {
+    data class Success<out T : Any>(val response: HttpResponse, val body: T) : NetworkResponse<T>()
+    data class Error<out E : KaliumException>(val kException: KaliumException) : NetworkResponse<E>()
+}
+
+fun <T> NetworkResponse<T>.successValue(): T = (this as NetworkResponse.Success).body
+fun <T> NetworkResponse<T>.isSuccessful(): Boolean = this is NetworkResponse.Success
```

Comment:
Amazing opportunity to use Contracts:

```suggestion

@OptIn(ExperimentalContracts::class)
fun <T> NetworkResponse<T>.isSuccessful(): Boolean {
    contract {
        returns(true) implies (this@isSuccessful is NetworkResponse.Success)
        returns(false) implies (this@isSuccessful is NetworkResponse.Error)
    }
    return this@isSuccessful is NetworkResponse.Success
}

```

This allows the compiler to infer the type and unlocks amazing auto-completion in the IDE:

```kotlin
        if(response.isSuccessful()){
            response.body //Because of the contract, the compiler already knows that this must be a success 
        }else{
            response.kException //Because of the contract, the compiler already knows that this must be an error
        }
```

Reply:
If we use this, maybe we could remove the `successValue()` and `errorValue()` casting functions. For quickly prototyping these functions are really useful, but they are a shortcut for mistakes, casting without knowing what's inside, and in any production application we would need to check every API call anyway.

Reply:
That was one of he inconveniences of using this kind of approach vs parsing manually the ktor result. However I LOVE these Contracts. I was not aware of their existance till now 🙏

Reply:
Now I'm wondering / playing around on how to integrate them when calling also `successValue()` and `errorValue()`

Reply:
I think something like this can work, but I don't see much value:

```kotlin

fun <T> NetworkResponse<T>.successValue(): Boolean {
    contract {
        returns() implies (this@isSuccessful is NetworkResponse.Success)
    }
    return this as NetworkResponse.Success
}


fun <T> NetworkResponse<T>.errorvalue(): Boolean {
    contract {
        returns() implies (this@isSuccessful is NetworkResponse.Error)
    }
    return this as NetworkResponse.Error
}
```

Basically: if it returns, it implies. The compiler will know about what it means, etc.
But if it doesn't return: 
# 💥 Boom

So we still need to check the value anyway... I think removing `successValueI()` or `errorValue()` for now isn't that bad.

`isSuccessful()` delivers a lot of value now and it's probably for the best that we don't provide unsafe casts in core functions of our library.

Reply:
The only way to prevent an `IllegalCastException` is by testing it. 

`errorValue()` and `successValue()` are not tests, they are assertions. So it will crash if it's not what the user expects.

`isSuccessful()` is one way of testing it. And by adding the contracts, the casts are done automatically by the compiler.

Instead of manually checking the type:
```kotlin
if(response is NetworkResponse.Success){ 
    showDataToUser(response.body) // Compiler knows about what this is
} else { 
    logAndHandleException(response.kException) 
}
```

We can make it like this:
```kotlin
if (response.isSuccessful()) {
    showDataToUser(response.body) // Compiler knows about what this is too, even though it's hidden behind another function
} else {
    logAndHandleException(response.kException)
}
```

Reply:
Hmm, I'm not entirely sold about the idea of adding yet another layer of checks. If we want to make sure calling `successValue/errorValue` will not throw yet another Exception (which was one of the main motivations for this change), I would suggest then to declare the return type as nullable `fun <T> NetworkResponse<T>.successValue(): T?`. 
That way the user will need to make sure of the validity of value returned before using the body of the response.